Support for reconnection

DemiMarie opened the pull request:
Plus a bunch of other changes, like ripping out the vendored copy of the libc crate.  See individual commit messages for details.

DemiMarie committed:
Recreate the vchan where needed

This is necessary to support GUI daemon reconnection.  To support this,
two new states (‘ReadState::Connecting’ and ‘ReadState::ReadingXConf’)
are added to the state machine.

DemiMarie committed:
Support reconnecting

This is necessary for the GUI agent.

DemiMarie committed:
Fix mis-remembered XConf size

Use the correct size, not 2.

DemiMarie committed:
Clear the queue when reconnecting

Otherwise the random stale data in it gets interpreted as GUI protocol
messages.

DemiMarie committed:
Set offset to 0 when resetting state

This avoids an assert.

DemiMarie committed:
Correctly implement reconnection

DemiMarie committed:
Rip out the old vendored copy of libc

Use the system packages from Fedora and Debian instead.

DemiMarie committed:
Document the meaning of an Ok(None) return value

DaemonToAgentEvent::parse() can return Ok(None) if the message number is
not known.

DemiMarie committed:
Add TryFrom impls for enum_const! enums

This will be very useful in other code.

DemiMarie committed:
Replace consts with enum variants

This was suggested by Jakub Kądziołka.

DemiMarie committed:
Remove unneeded imports

This was suggested by Jakub Kądziołka.

DemiMarie committed:
Use an associated const instead of a function

This was suggested by Jakub Kądziołka.

marmarek commented:
@NieDzejkob can you verify if that's okay now? and also review the newly added code here

DemiMarie committed:
Use more idiomatic conversion function

TryInto is more idiomatic than multiple casts.

This was suggested by Jakub Kądziołka.

DemiMarie committed:
Rename `set` to `present_windows`

The latter name is more descriptive.

This was suggested by Jakub Kądziołka.

DemiMarie committed:
Rename buffer::Vchan to RawMessageStream

This avoids confusion with the raw Vchan type.  Suggested by Jakub
Kądziołka.

DemiMarie committed:
Rename drain() to flush_pending_writes()

This is more descriptive.  Suggested by Jakub Kądziołka.

DemiMarie committed:
Support creating a Castable object from a byte slice

This is done using MaybeUninit and copy_nonoverlapping.

DemiMarie committed:
Add more tests for from_bytes

The function uses unsafe code internally, so it is critical to test it
thoroughly.  Thankfully, it has only one branch to check.

DemiMarie committed:
Add a `Castable::zeroed()` helper

This provides a zeroed instance of any Castable type.

DemiMarie committed:
Use ptr::read_unaligned() to implement from_bytes

This is simpler than using ptr::copy_nonoverlapping, but may need a
special case for zero-sized types.  It is simpler to just special-case
zero-sized types then check if it is necessary to do so.  The branch
will be constant-folded by the compiler and has no run-time overhead.

DemiMarie committed:
Inline all Castable functions and methods

It makes no sense to *not* inline any of these functions.

DemiMarie committed:
Document that Castable types are never Drop

This means they do not have destructors that need to be called for them.

DemiMarie committed:
Use Castable::from_bytes in qubes-gui-agent-proto

This allows removing a large amount of boilerplate code.
